src/libostree/ostree-repo-pull.c

```diff
@@ -1466,6 +1470,167 @@ process_verify_result (OtPullData            *pull_data,
+        loaded_from_file = TRUE;
+      else
+          g_debug("Unable to load public keys for '%s' from file '%s': %s",
+                  ostree_sign_get_name(sign), pk_file, (*error)->message);
+    }
+
+  if (pk_ascii != NULL)
+    {
+      g_autoptr (GError) local_error = NULL;
+      g_autoptr (GVariant) pk = g_variant_new_string(pk_ascii);
+
+      /* Add inlined public key */
+      if (loaded_from_file)
+        loaded_inlined = ostree_sign_add_pk (sign, pk, &local_error);
+      else
+        loaded_inlined = ostree_sign_set_pk (sign, pk, &local_error);
+
+      if (!loaded_inlined)
+        {
+          g_debug("Unable to load public key '%s' for '%s': %s",
+                  pk_ascii, ostree_sign_get_name(sign), local_error->message);
+
+          /* Save error message for better reason detection later if needed */
+          if (*error == NULL)
+            g_propagate_error (error, g_steal_pointer (&local_error));
+          else
+            g_prefix_error (error, "%s; ", local_error->message);
+        }
+    }
+
+  /* Return true if able to load from any source */
+  if (loaded_from_file || loaded_inlined)
+    {
+      g_clear_error (error);
+      return TRUE;
+    }
+
+  return FALSE;
+}
+
+static gboolean
+_ostree_repo_sign_verify (OstreeRepo *repo,
+                          const gchar *remote_name,
+                          GBytes *signed_data,
+                          GVariant *metadata,
+                          GError **error)
+{
+  /* list all signature types in detached metadata and check if signed by any? */
+  g_auto (GStrv) names = ostree_sign_list_names();
+  for (char **iter=names; iter && *iter; iter++)
+    {
+      g_autoptr (OstreeSign) sign = NULL;
+      g_autoptr (GVariant) signatures = NULL;
+      const gchar *signature_key = NULL;
+      GVariantType *signature_format = NULL;
+      g_autoptr (GError) local_error = NULL;
+
+      if ((sign = ostree_sign_get_by_name (*iter, &local_error)) == NULL)
+        continue;
+
+      signature_key = ostree_sign_metadata_key (sign);
+      signature_format = (GVariantType *) ostree_sign_metadata_format (sign);
+
+      signatures = g_variant_lookup_value (metadata,
+                                           signature_key,
+                                           signature_format);
+
+      /* If not found signatures for requested signature subsystem */
+      if (!signatures)
+        continue;
+
+      /* Try to load public key(s) according remote's configuration */
+      if (_load_public_keys (sign, repo, remote_name, &local_error))
+        {
+          /* Return true if any signature fit to pre-loaded public keys.
+           * If no keys configured -- then system configuration will be used */
+          if (ostree_sign_data_verify (sign,
+                                       signed_data,
+                                       signatures,
+                                       &local_error))
+            {
+              g_clear_error (error);
```

Comment:
Similarly here I think it's best if we hoist the `local_error` out of the loop here; call it something like `last_error`, and the loop shouldn't be touching `error`.

At the end of the loop we just propagate the error if one is set, or return the "no supported signed type".